File: app/src/main/java/com/project/poupay/view/ListSegmentAdapter.java
package com.project.poupay.view;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.project.poupay.R;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

public class ListSegmentAdapter extends RecyclerView.Adapter<ListSegmentAdapter.ViewHolder> {

    private final LinkedHashMap<String, List<ContentItem>> mItems;

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView titleText, totalText;
        private final LinearLayout content;

        public ViewHolder(View view) {
            super(view);
            titleText = view.findViewById(R.id.Segment_Title);
            totalText = view.findViewById(R.id.Segment_Total);
            content = view.findViewById(R.id.Segment_Content);
        }

        public void setup(String title, List<ContentItem> items) {
            double total = 0;
            titleText.setText(title);
            content.removeAllViews();
            for (ContentItem item : items) {
                content.addView(item);
                total += item.getValue();
            }
            totalText.setText(NumberFormat.getCurrencyInstance(Locale.getDefault()).format(total));
        }
    }

    public ListSegmentAdapter() {
        mItems = new LinkedHashMap<>();
    }

    public void add(ContentItem newItem) {
        String key = newItem.getDate();
        if (mItems.containsKey(key)) {
            Objects.requireNonNull(mItems.get(key)).add(newItem);
        } else {
            List<ContentItem> list = new ArrayList<>();
            list.add(newItem);
            mItems.put(key, list);
        }
        notifyItemInserted(getItemCount() - 1);
    }

    public void add(List<ContentItem> newItems) {
        for (ContentItem item : newItems) add(item);
    }

    public void clear() {
        int count = getItemCount();
        mItems.clear();
        notifyItemRangeRemoved(0, count);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.view_segment, viewGroup, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, final int position) {
        String title = (String) mItems.keySet().toArray()[position];
        List<ContentItem> items = new ArrayList<>(Objects.requireNonNull(mItems.get(title)));
        viewHolder.setup(title, items);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }



}

File: app/src/main/java/com/projeto/poupay/view/ListSegmentAdapter.kt
package com.projeto.poupay.view

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.projeto.poupay.R
import java.text.NumberFormat
import java.util.Locale

class ListSegmentAdapter : RecyclerView.Adapter<ListSegmentAdapter.ViewHolder>() {
    private var mItems: LinkedHashMap<String, MutableList<ContentItem>> = LinkedHashMap()

    class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
        private val titleText: TextView
        private val totalText: TextView
        private val content: LinearLayout

        init {
            with(view) {
                titleText = findViewById(R.id.Segment_Title)
                totalText = findViewById(R.id.Segment_Total)
                content = findViewById(R.id.Segment_Content)
            }
        }

        fun setup(title: String, items: List<ContentItem>) {
            var total = 0.0
            titleText.text = title
            content.removeAllViews()
            for (item in items) {
                content.addView(item)
                total += item.value
            }
            totalText.text = NumberFormat.getCurrencyInstance(Locale.getDefault()).format(total)
        }
    }

    fun add(newItem: ContentItem) {
        val key = newItem.date
        if (mItems.containsKey(key)) {
            mItems[key]?.add(newItem)
        } else {
            val list = mutableListOf<ContentItem>()
            list.add(newItem)
            mItems[key] = list
        }
        notifyItemInserted(itemCount - 1)
    }

    fun add(newItems: List<ContentItem>) {
        for (newItem in newItems) {
            add(newItem)
        }
    }

    fun clear() {
        val count = itemCount
        mItems.clear()
        notifyItemRangeRemoved(0, count)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.view_segment, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val title: String = mItems.keys.toList()[position]
        val items: List<ContentItem> = mItems[title] ?: listOf()
        holder.setup(title, items)
    }
}